Updates to scRNA filtering (#249)
AlexsLemonade/training-modules: scRNA-seq/01-filtering_scRNA-seq.Rmd

---

> title: "scRNA-seq Normalization"

Reviewer: This title needs to be updated.

Reviewer: Sorry this was supposed to be a review comment!

output:   
  html_notebook: 
    toc: true
    toc_float: true
---

**CCDL 2019**

In this notebook, we'll perform quality control analyses of scRNA-seq count data. 

> The [data set] (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84465) we are using is cells from a glioblastoma sample  that were Fluorescence-Activated Cell sorted (FACS) and then processed by paired end sequencing of RNA using the [Smart-seq2 protocol](http://www.nature.com/doifinder/10.1038/nmeth.2639) 

Reviewer: ```suggestion
The [data set](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84465) we are using is cells from a glioblastoma sample that were Fluorescence-Activated Cell sorted (FACS) and then processed by paired end sequencing of RNA using the [Smart-seq2 protocol](http://www.nature.com/doifinder/10.1038/nmeth.2639) 
```

                        total_counts,
                        num_genes_exp,
                        mito_fraction) %>%
  # add on the cell type again
  dplyr::left_join(dplyr::select(sc_metadata, 
                                 geo_accession, cell_type = cell.type.ch1)) %>%
  dplyr::arrange(mito_fraction) # sort for plotting order
  
```
Now we can plot these measures all together, along with some possible cutoffs

```{r}
ggplot(qc_df, aes (x = total_counts,
                   y = num_genes_exp, 
                   color = mito_fraction)) +
  geom_point(alpha = 0.5) +
  scale_color_viridis_c() + 
  geom_vline(xintercept = 300000, color = "red") +
  geom_hline(yintercept = 4000, color = "red") +
  labs(x = "Total Count",
       y = "Number of Genes Expressed",
       color = "Mitochondrial\nFraction") + 
  theme_bw()
```

Now if we want to filter our data based on these measures and cutoffs we like, we can do this with `dplyr::filter` and then select the resulting columns from the matrix.

```{r}
filtered_samples <- qc_df %>%
  dplyr::filter(total_counts > 300000,
                num_genes_exp > 4000,
                mito_fraction < 0.1)

> sc_data_filtered <- sc_data[,filtered_samples$geo_accession]

Reviewer: ```suggestion
sc_data_filtered <- sc_data[, filtered_samples$geo_accession]
```

Now we have an idea of what samples we probably want to get rid of.
But what if our data contains genes that we can't reliably measure in these cells?

Let's use our `detection_mat` to add up how many samples express each of 
these genes so we can try to filter out unreliably measured genes.  

```{r}
# Make a vector that contains the number of cells that express a particular gene
num_samples_that_exp <- rowSums(detection_mat)
```

Let's make another density plot with the number of samples that express each gene:

```{r}
# Let's take a look at what this looks like:
ggplot(mapping = aes(x = num_samples_that_exp)) +
  geom_density(fill = "lightblue") +
  geom_vline(xintercept = 200, col = "red") +
  xlab("Number of Samples Expressing Each Gene") + 
  theme_classic()
```

How many genes will be excluded if we draw our cutoff at 200 samples?

```{r}
sum(num_samples_that_exp < 200)
```
That's a lot! How do we feel about that?

To filter the genes that are not expressed in many samples, we will take a somewhat simpler approach, and use base R to select the rows from our matrix (already filtered by sample) that are above the cutoff.

```{r}

> cutoff = 200

Reviewer: ```suggestion
cutoff <- 200
```

---
title: "Single cell RNA-seq quality control and filtering"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

> **CCDL 2019**

Reviewer 2: ```suggestion
**CCDL 2020**
```

In this notebook, we'll perform quality control analyses of scRNA-seq count data. 
The [data set](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84465) we are using is cells from a glioblastoma sample that were Fluorescence-Activated Cell sorted (FACS) and then processed by paired end sequencing of RNA using the [Smart-seq2 protocol](http://www.nature.com/doifinder/10.1038/nmeth.2639) 
[(Darmanis *et al.* 2017)](https://www.ncbi.nlm.nih.gov/pubmed/29091775).

  
## Set Up 

```{r setup}
# Magrittr for the pipe %>%
library(magrittr)

# GGPlot2 for the plots
library(ggplot2)
```

## Import single-cell RNA-seq counts matrix

The Smart-seq2 sequencing results in separate paired fastq files for each processed cell (in this case over 7000 total fastq files).
We have processed this data with `salmon quant` and used `tximport` to create a gene-level count matrix, filtering samples to only include those with at least 30% of reads mapping.

![](diagrams/full-length_1.png) 

> We will discuss *tag-based* scRNA-seq data (such as 10X), in a later notebook,  where you will learn how to process raw data to a count matrix and could apply these same types of steps to filter and normalize the data.

Reviewer 2: ```suggestion
We will discuss *tag-based* scRNA-seq data (such as 10X), in a later notebook, where you will learn how to process raw data to a count matrix and could apply these same types of steps to filter and normalize the data.
```

### Directories and files

The data files we will be using for this part of the project are in the `data/glioblastoma` subdirectory of the `scRNA-seq` directory where this file is located.

> The main files are further located in the `preprocessed` subdirectory (and sometimes deeper than that!). 

Reviewer 2: ```suggestion
The main files are further located in the `preprocessed` subdirectory. 
```
Makes it sound like the files might switch places? (Maybe I'm overthinking it). Keep it if you like.

Author: I'm going to modify this to be a bit more clear, but keep the sentiment.

A figure of the (partial) directory structure for this module is shown below.

![](diagrams/glioblastoma_dir_structure.png)

Importantly, the `preprocessed` directory is actually a *link* (think of a shortcut or alias) to a shared folder, so everyone is looking at exactly the same sets of files in there, and you will not be able to write to that directory.

> This is actually a good idea/practice to enforce in your own projects, if you can.

Reviewer 2: Combining these two sentences to get to your point would be nice. 
```suggestion
It's good practice to store your data in a place that you should never write to, except when adding new data.
```

Author: Yeah, I like this.

Recall that we can do this because all the remaining data has the same type.
While not strictly necessary, having the data in the form of a matrix will make some calculations more efficient.

```{r make_matrix}
# Set the column as the gene names
sc_data <- tibble::column_to_rownames(sc_data, "gene")
sc_data <- as.matrix(sc_data)
```

Let's look at the mean expression of the genes in this dataset. 
We will use `apply()` in order to calculate things across our data frame. 
The second argument in `apply()` specifies whether we are calculating by rows or columns. 
(1 = rows, 2 = columns).

In the code chunk below, use `apply()` with the correct arguments to calculate the gene means. 

```{r means, live = TRUE}
# Let's calculate the gene means (by row)
gene_means <- apply(sc_data, 1, mean)
```

This works just fine, but you may have noticed it is a bit slow.
For a few common summary functions like means and sums, R has much more efficient functions to calculate across rows or columns. 
In this case, we can use `rowMeans()` to do the same calculation, much more quickly.

```{r rowmeans}
# use rowMeans() to calculate gene means
gene_means <- rowMeans(sc_data)
```

> 

Reviewer 2: ```suggestion
```
I don't like the extra spacings we have in these notebooks and not the others. It really doesn't matter, I'm being nit picky, but I needed to say it.

  dplyr::select(geo_accession, num_genes_exp, cell_type = cell.type.ch1)

```

The next graph we are using the same steps as before, but to make individual graphs for each cell type, we are using the ggplot2 `facet_wrap()` function.

```{r facet_genes, live = TRUE}
# Plot with faceting by cell type
ggplot(num_genes_exp_df, aes(x = num_genes_exp)) + 
  xlab("Number of Genes Expressed") +
  geom_density(fill = "lightblue") + 
  geom_vline(xintercept = 3000, color = "red") + 
  facet_wrap(~ cell_type) + # Facet wrap uses labels you give it to make individual graphs. 
  theme_classic()
```

Now we are aware of how our filtering may be affecting different cell types differently. 

### Mitochondrial gene expression

If a cell is dead or dying, its mRNA will tend to leak out of the cell, leaving an overabundance of mitochondrial RNA, which is more likely to stay within the mitochondria longer. 
To look for this, we would like to calculate the fraction of mitochondrial expression for each cell as well.
First, we will need a list of the mitochondrial genes, which we have prepared in a tsv file `hs_mitochondrial_genes.tsv` that we will now read in:

```{r read_mito}

> #read `hs_mitochondrial_genes.tsv` from data_dir

Reviewer 2: ```suggestion
# read `hs_mitochondrial_genes.tsv` from data_dir
```

# sum the counts from just those genes for all samples
mito_counts <- colSums(sc_data[mito_rows, ])

# calculate mito_fraction for all samples
mito_fraction <- mito_counts/total_counts
```

Lets make a plot of this distribution as well!

```{r plot_mito, live = TRUE}
ggplot(mapping = aes(x = mito_fraction)) +
  geom_density(fill = "lightblue") + 
  xlab("Mitchondrial fraction") +
  theme_classic()
```

### Combining sample QC measures

Lets put all of the QC measures we have calculated into a single data frame, so we can look at how they might relate to one another.

```{r qc_dataframe, live = TRUE}
# create a data frame of 
# geo_accession, total_counts, num_genes_exp, and mito_fraction
qc_df <- tibble::tibble(geo_accession = names(total_counts),
                        total_counts,
                        num_genes_exp,
                        mito_fraction) %>%
  # add on the cell type from metadata with left_join()
  dplyr::left_join(dplyr::select(sc_metadata, 
                                 geo_accession, cell_type = cell.type.ch1)) %>%
  # sort by mito_fraction for plotting 
  dplyr::arrange(mito_fraction) 
  
```

> Now we can plot these measures all together, along with some possible cutoffs

Reviewer 2: ```suggestion
Now we can plot these measures all together, along with some possible cutoffs.
```

```{r qc_scatterplot}
ggplot(qc_df, aes (x = total_counts,
                   y = num_genes_exp, 
                   color = mito_fraction)) +
  geom_point(alpha = 0.5) +
  scale_color_viridis_c() + 
  geom_vline(xintercept = 300000, color = "red") +
  geom_hline(yintercept = 4000, color = "red") +
  labs(x = "Total Count",
       y = "Number of Genes Expressed",
       color = "Mitochondrial\nFraction") + 
  theme_bw()
```

> If we want to filter our data based on these measures and cutoffs we like, we can do this with `dplyr::filter` and then select the resulting columns from the matrix.

Reviewer 2: ```suggestion
If we want to filter our data based on these measures and cutoffs we like, we can do this with `dplyr::filter()` and then select the resulting columns from the matrix.
```

Let's use our `detection_mat` to add up how many samples express each of these genes so we can try to filter out unreliably measured genes.  

```{r sample_exp, live = TRUE}
# Make a num_samples_exp vector that contains the 
# number of cells that express a particular gene
num_samples_exp <- rowSums(detection_mat)
```

Let's make another density plot with the number of samples that express each gene:

```{r sample_exp_plot}
# Let's take a look at what this looks like:
ggplot(mapping = aes(x = num_samples_exp)) +
  geom_density(fill = "lightblue") +
  geom_vline(xintercept = 200, col = "red") +
  xlab("Number of Samples Expressing Each Gene") + 
  theme_classic()
```

How many genes will be excluded if we draw our cutoff at 200 samples?

```{r filter_effect, live = TRUE}
sum(num_samples_exp < 200)
```
That's a lot! How do we feel about that?

To filter the genes that are not expressed in many samples, we will take a somewhat simpler approach, and use base R to select the rows from our matrix (already filtered by sample) that are above the cutoff.

```{r filter_genes}
cutoff <- 200
# filter sc_data_filtered to only genes above cutoff value
# save as sc_data_filtered2

> sc_data_filtered2 <- sc_data_filtered[num_samples_exp >= cutoff, ]

Reviewer 2: ```suggestion
sc_data_filtered_hi_exp <- sc_data_filtered[num_samples_exp >= cutoff, ]
```
Can we name it something else? I feel like participants will take this as license to start naming things tmp1, tmp2, tmp3.

Author: You are of course correct here, but I am going to choose a different name. `sc_data_filtered_gene_express`, I think.

  theme_classic()
```

Now we are aware of how our filtering may be affecting different cell types differently. 

### Mitochondrial gene expression

If a cell is dead or dying, its mRNA will tend to leak out of the cell, leaving an overabundance of mitochondrial RNA, which is more likely to stay within the mitochondria longer. 
To look for this, we would like to calculate the fraction of mitochondrial expression for each cell as well.
First, we will need a list of the mitochondrial genes, which we have prepared in a tsv file `hs_mitochondrial_genes.tsv` that we will now read in:

```{r read_mito}
#read `hs_mitochondrial_genes.tsv` from data_dir
mito_genes_df <- readr::read_tsv(mito_genes_file) 
```

Now we can use the `gene_id`s from that list to select only the rows of the count matrix that correspond to the mitochondrial genes and sum their expression for each sample.

```{r mito_filter, live = TRUE}
# create a mito_rows vector that is TRUE for mitochondrial genes in our dataset
mito_rows <- rownames(sc_data) %in% mito_genes_df$gene_id

# sum the counts from just those genes for all samples
mito_counts <- colSums(sc_data[mito_rows, ])

# calculate mito_fraction for all samples

> mito_fraction <- mito_counts/total_counts

Author: Yeah, I was just debating when to introduce this; I was thinking I might show that we can do all the same QC from within *SingleCellExperiment* when using the tag data, but I thought showing it all with base R/tidyverse for an introduction might still be a good way to go.
